src/app/products/product-detail.guard.ts

```diff
@@ -0,0 +1,21 @@
+import { Injectable } from '@angular/core';
+import { CanActivate, ActivatedRouteSnapshot, Router } from '@angular/router';
+import { Observable } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProductDetailGuard implements CanActivate {
+
+  constructor(private router: Router) { }
+
+  canActivate(activatedRouteNext: ActivatedRouteSnapshot): Observable<boolean> | Promise<boolean> | boolean {
+    const id = +activatedRouteNext.url[1].path;
+    if (isNaN(id) || id < 1) {
+      alert('Invalid product Id');
```

Comment:
Please Remove the alert.

Reply:
Removed alert. Will add Toast message instead of this in the next PR